daemontools-x/svstat.8
.\" vim: tw=75
.TH svstat 8
.SH NAME
svstat \- print status of services monitored by \fBsupervise\fR(8).

.SH SYNOPSIS
\fBsvstat\fR \fIservices\fR

.SH DESCRIPTION
\fIservices\fR consists of any number of arguments, each argument naming a
directory. \fBsvstat\fR prints one human-readable line for each directory,
saying whether \fBsupervise\fR(8) is successfully running in that
directory, and reporting the status and process information maintained by
\fBsupervise\fR. The information printed is the \fBstatus\fR and \fBpid\fR.
The \fBstatus\fR is one of \fIup\fR, \fIdown\fR or \fIwait\fR. In case the
service status is \fIup\fR, \fBpid\fR is the process id of the service. In
case the status is either \fIdown\fR or \fIwait\fR, \fBpid\fR will be the
process id of the \fBsupervise\fR process.

When \fBsupervise\fR is monitoring a process that forks in the background,
the pid that is shown may be actually no longer exist. In such a case
\fBsvstat\fR will show pid label as '?pid'. As stated in the
\fBsupervise\fR man page, you can have \fBsupervise\fR monitor such errant
processes by using \fBfghack\fR(8) or by having sticky bit on the run file.

.EX
# this service is up. here the pid of qmail-qmtpd is 762703
$ svstat /service/qmail-qmtpd.209
/service/qmail-qmtpd.209: up 64849.402144999 seconds pid 762703

# this service is down. here the pid of supervise is 762596
$ svstat /service/qmail-qmqpd.628
/service/qmail-qmqpd.628: down 3453.701675999 seconds spid 762596

# this service is waiting for a dependent service to be up
# It has been waiting for the past 96 seconds. The pid
# 762630 is the pid of the supervise process
$ svstat /service/inlookup.infifo
/service/inlookup.infifo: wait 96.530758999 seconds spid 762630

# this service has already waited 3.28 seconds since a dependent
# service came up. It will further wait for 7 seconds. The pid
# 762630 is the pid of the supervise process
$ svstat /service/inlookup.infifo
/service/inlookup.infifo: wait 3.285475999 seconds spid 762630 remaining 7 seconds
.EE

If the argument to \fBsvscan\fR names just one directory, then the return
value will reflect various status of a service as below

.EX
	 111 - system error
	 100 - bad status
	   0 - service running
	   1 - service down
	   2 - supervise not running
	   3 - service paused
	   4 - down - wants up
	   5 - up   - wants down
	   6 - wait - wating for service to come up
.EE

.SH SEE ALSO
supervise(8),
svps(1),
svc(8),
svok(8),
svwait(8),
svscan(8),
svctool(8),
minisvc(8),
svscanboot(8),
readproctitle(8),
fghack(8),  
pgrphack(8),
multilog(8),
tai64n(8),
tai64nlocal(8),
setuidgid(8),
envuidgid(8),
envdir(8),
softlimit(8),
setlock(8)

http://cr.yp.to/daemontools.html


libdkim-x/NEWS
Migrated to libtool 2.x


daemontools-x/svstat.c
/*
 * $Id: svstat.c,v 1.15 2026-01-26 20:13:41+05:30 Cprogrammer Exp mbhangui $
 */
#include <unistd.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/stat.h>
#include "strerr.h"
#include "error.h"
#include "open.h"
#include "fmt.h"
#include "taia.h"
#include "substdio.h"
#ifdef USE_RUNFS
#include "run_init.h"
#endif

#define FATAL "svstat: fatal: "
#define WARN  "svstat: warning: "

char            outbuf[256], errbuf[256];
substdio        o = SUBSTDIO_FDBUF((ssize_t (*)(int,  char *, size_t)) write, 1, outbuf, sizeof outbuf);

char            status[21];
char            strnum[FMT_DOUBLE];

unsigned long   pid;
unsigned char   normallyup, want, paused;
static short    waiting;

void
doit(char *dir, int *retval)
{
	struct stat     st;
	int             r, fd = -1;
	short          *s;
	const char     *x;
	struct taia     when, now;

	*retval = 111;
	if (chdir(dir) == -1) {
		strerr_warn4(WARN, "unable to to chdir to ", dir, ": ", &strerr_sys);
		return;
	}

	normallyup = 0;
	if (stat("down", &st) == -1) {
		if (errno != error_noent) {
			strerr_warn4(WARN, "unable to stat ", dir, "/down: ", &strerr_sys);
			return;
		}
		normallyup = 1;
	}

#ifdef USE_RUNFS
	if ((fd = open_write("supervise/ok")) == -1) {
		if (errno == error_nodevice) {
			strerr_warn4(WARN, "unable to open ", dir, "/supervise/ok: supervise not running: ", &strerr_sys);
			*retval = 2;
			return;
		} else
		if (errno == error_noent)
			*retval = 2;
		else {
			strerr_warn4(WARN, "unable to open ", dir, "/supervise/ok: ", &strerr_sys);
			return;
		}
		switch (run_init(dir))
		{
		case 0: /*- cwd changed to /run/svscan/service_name */
			break;
		case 1: /*- /run, /var/run doesn't exist */
			strerr_warn4(WARN, "unable to open ", dir, "/supervise/ok: supervise not running: ", &strerr_sys);
		case -1:
			strerr_warn2(WARN, "unable to get current working directory: ", &strerr_sys);
			return;
		case -2:
			strerr_warn4(WARN, "No run state information for ", dir, ": ", &strerr_sys);
			return;
		case -3:
			strerr_warn3(WARN, dir, ": name too long", 0);
			return;
		}
	}
#endif
	if (fd == -1 && (fd = open_write("supervise/ok")) == -1) {
		if (errno == error_nodevice || errno == error_noent) {
			strerr_warn4(WARN, "unable to open ", dir, "/supervise/ok: supervise not running: ", &strerr_sys);
			*retval = 2;
			return;
		}
		strerr_warn4(WARN, "unable to open ", dir, "/supervise/ok: ", &strerr_sys);
		return;
	}
	close(fd);
	if ((fd = open_read("supervise/status")) == -1) {
		if (errno == error_noent)
			*retval = 2;
		strerr_warn4(WARN, "unable to open ", dir, "/supervise/status: ", &strerr_sys);
		return;
	}
	r = read(fd, status, sizeof status);
	close(fd);
	if (r < sizeof status) {
		if (r == -1)
			x = error_str(errno);
		else {
			*retval = 100;
			x = "bad format";
		}
		strerr_warn5(WARN, "unable to read ", dir, "/supervise/status: ", x, 0);
		return;
	}
	pid = (unsigned char) status[15];
	pid <<= 8;
	pid += (unsigned char) status[14];
	pid <<= 8;
	pid += (unsigned char) status[13];
	pid <<= 8;
	pid += (unsigned char) status[12];
	if ((paused = status[16]))
		*retval = 3;
	want = status[17];
	s = (short *) (status + 18);
	waiting = *s;
	taia_unpack(status, &when);
	taia_now(&now);
	if (taia_less(&now, &when))
		when = now;
	taia_sub(&when, &now, &when);
	substdio_puts(&o, dir);
	substdio_puts(&o, ": ");
	if (waiting)
		substdio_puts(&o, "wait ");
	else
	if (status[20])
		substdio_puts(&o, "up ");
	else
		substdio_puts(&o, "down ");
	substdio_put(&o, strnum, fmt_double(strnum, when.sec.x, 0));
	substdio_put(&o, strnum, fmt_ulong(strnum, when.nano));
	substdio_puts(&o, " seconds");

	if (status[20] && !normallyup)
		substdio_puts(&o, ", normally down");
	if (!pid && normallyup)
		substdio_puts(&o, ", normally up");
	if (status[20] && paused)
		substdio_puts(&o, ", paused");
	if (!pid && (want == 'u')) {
		*retval = 4;
		substdio_puts(&o, ", want up");
	}
	if (status[20] && (want == 'd')) {
		*retval = 5;
		substdio_puts(&o, ", want down");
	}

	if (pid && status[20]) {
		if (*retval != 3 && *retval != 5)
			*retval = 0;
		if (kill(pid, 0))
			substdio_puts(&o, " ?pid ");
		else
			substdio_puts(&o, "  pid ");
		substdio_put(&o, strnum, fmt_ulong(strnum, pid));
		substdio_puts(&o, " ");
	} else
	if (pid) {
		*retval = 1;
		substdio_puts(&o, " spid ");
		substdio_put(&o, strnum, fmt_ulong(strnum, pid));
		substdio_puts(&o, " ");
	} else
		*retval = 1;
	if (waiting > 0 && (waiting - when.sec.x) > 0) {
		substdio_puts(&o, "remaining ");
		when.sec.x = waiting - when.sec.x;
		substdio_put(&o, strnum, fmt_ulong(strnum, tai_approx(&when.sec)));
		substdio_puts(&o, " seconds");
	}
	substdio_puts(&o, "\n");
}

int
main(int argc, char **argv)
{
	int             fdorigdir, retval;
	char           *dir;

	++argv;

	if ((fdorigdir = open_read(".")) == -1)
		strerr_die2sys(111, FATAL, "unable to open current directory: ");
	while ((dir = *argv++)) {
		doit(dir, &retval);
		if (fchdir(fdorigdir) == -1)
			strerr_die2sys(111, FATAL, "unable to revert directory: ");
	}
	substdio_flush(&o);
	/*
	 * 111 - system error
	 * 100 - bad status
	 *   0 - service running
	 *   1 - service down
	 *   2 - supervise not running
	 *   3 - service paused
	 *   4 - down - wants up
	 *   5 - up   - wants down
	 *   6 - wait - wating for service to come up
	 */
	_exit(retval);
}

void
getversion_svstat_c()
{
	const char     *x = "$Id: svstat.c,v 1.15 2026-01-26 20:13:41+05:30 Cprogrammer Exp mbhangui $";

	x++;
}

/*
 * $Log: svstat.c,v $
 * Revision 1.15  2026-01-26 20:13:41+05:30  Cprogrammer
 * display ?pid when process being supervised forks into background
 *
 * Revision 1.14  2026-01-11 20:36:24+05:30  Cprogrammer
 * report uptime in seconds.nanoseconds
 *
 * Revision 1.13  2025-01-21 23:35:45+05:30  Cprogrammer
 * Fixes for gcc14
 *
 * Revision 1.12  2024-05-09 22:39:36+05:30  mbhangui
 * fix discarded-qualifier compiler warnings
 *
 * Revision 1.11  2023-03-04 14:43:01+05:30  Cprogrammer
 * check for supervise/ok in original service dir before run filesystem
 *
 * Revision 1.10  2020-11-30 23:20:31+05:30  Cprogrammer
 * change warning text message for missing directory in /run
 *
 * Revision 1.9  2020-11-30 22:56:02+05:30  Cprogrammer
 * continue instead of exit on run_init() failure
 *
 * Revision 1.8  2020-11-10 19:13:45+05:30  Cprogrammer
 * use byte 20 from status to indicate if service is up
 *
 * Revision 1.7  2020-11-09 09:31:09+05:30  Cprogrammer
 * display wait status. print errors to stderr instead of stdout
 *
 * Revision 1.6  2020-11-07 14:23:49+05:30  Cprogrammer
 * print pid after displaying uptime
 *
 * Revision 1.5  2020-10-08 18:34:17+05:30  Cprogrammer
 * use /run, /var/run if system supports it
 *
 * Revision 1.4  2020-09-27 14:16:09+05:30  Cprogrammer
 * display status written by svwait command
 *
 * Revision 1.3  2004-10-22 20:31:20+05:30  Cprogrammer
 * added RCS id
 *
 * Revision 1.2  2004-10-09 23:35:02+05:30  Cprogrammer
 * replaced buffer functions with substdio
 *
 * Revision 1.1  2003-12-31 18:36:54+05:30  Cprogrammer
 * Initial revision
 *
 */
